Framework/Algorithms/src/CreateUserDefinedBackground.cpp

```diff
@@ -0,0 +1,258 @@
+  sortArgs.emplace_back(table->getColumn(0)->name(), true);
+  table->sort(sortArgs);
+}
+
+/**
+ * Extend background to limits of data: if it doesn't extend this far already,
+ * insert first and last points of data into it.
+ * @param background :: [input, output] Table of background points to work on
+ * @param data :: [input] Input workspace with data
+ */
+void CreateUserDefinedBackground::extendBackgroundToData(
+    API::ITableWorkspace_sptr &background,
+    const API::MatrixWorkspace_const_sptr &data) const {
+  const auto &xPoints = data->points(0);
+
+  // If first point > data minimum, insert a new first point
+  if (background->Double(0, 0) > xPoints[0]) {
+    background->insertRow(0);
+    API::TableRow firstRow = background->getFirstRow();
+    firstRow << xPoints.front() << background->Double(0, 1);
+  }
+  // If last point < data maximum, append a new last point
+  const size_t endIndex = background->rowCount() - 1;
+  if (background->Double(endIndex, 0) < xPoints.back()) {
+    API::TableRow lastRow = background->appendRow();
+    lastRow << xPoints.back() << background->Double(endIndex, 1);
+  }
+}
+
+/**
+ * Given a table of background points and the original workspace,
+ * return a new workspace containing interpolated background data.
+ * The same background is assumed for all spectra.
+ * @param background :: [input] Table of background points
+ * @param data :: [input] Original data workspace
+ * @returns :: Workspace containing background to be subtracted
+ */
+API::MatrixWorkspace_sptr
+CreateUserDefinedBackground::createBackgroundWorkspace(
+    const API::ITableWorkspace_const_sptr &background,
+    const API::MatrixWorkspace_const_sptr &data) const {
+  auto outputWS = data->clone();
+
+  const auto &xPoints = outputWS->points(0);
+  std::vector<double> yBackground;
+  std::vector<double> eBackground(xPoints.size(), 0);
+
+  // Interpolate Y data in the table to get y for each point
+  const auto &lerp = getInterpolator(background, data);
+  for (const auto &x : xPoints) {
+    const double y = lerp.value(x);
+    yBackground.push_back(y);
+  }
+
+  // Find out if user has selected the option "normalize histogram to bin width"
+  const bool distributionPlot = []() {
+    const auto retval =
+        Kernel::ConfigService::Instance().getString(AUTODISTRIBUTIONKEY);
+    return retval == "On";
+  }();
+
+  // Y is frequencies for distribution data (histo), otherwise counts
+  const bool useFrequencies =
+      data->isHistogramData() && (data->isDistribution() || distributionPlot);
+
+  // Apply Y and E data to all spectra in the workspace
+  for (size_t spec = 0; spec < outputWS->getNumberHistograms(); spec++) {
+    if (useFrequencies) {
+      outputWS->setFrequencies(spec, yBackground);
+      outputWS->setFrequencyStandardDeviations(spec, eBackground);
+    } else {
+      outputWS->setCounts(spec, yBackground);
+      outputWS->setCountStandardDeviations(spec, eBackground);
```

Comment:
@tom-perkins I think this block is making trouble in the unit tests now. For ensuring correctness, setting `Counts` when we have frequency data is now forbidden (and vice versa). The problem is basically the `useFrequencies` flag. I don't know what the best solution here is maybe changing the type of `outputWS` is the right way?